File: Combinations/Combinations.java

import java.util.ArrayList;
import java.util.List;

public class Combinations {
    private static int n = 4;
    private static int k = 2;
    private static List<List<Integer>> output = new ArrayList<>();

    public static void main(String[] args) {
        backtrack(1, new ArrayList<>());
        System.out.println(output);
    }

    private static void backtrack(int first, List<Integer> curr) {
        if (curr.size() == k) {
            output.add(new ArrayList<>(curr));
            return;
        }
        for (int i = first; i <= n; i++) {
            curr.add(i);
            backtrack(i + 1, curr);
            curr.remove(curr.size() - 1); 
        }
    }
}

File: Combinations/Combinations.py
n=4
k=2


def backtrack(first=1, curr=[]):
    if len(curr) == k:
        output.append(curr[:])
        return
    for i in range(first, n + 1):
        curr.append(i)
        backtrack(i + 1, curr)
        curr.pop()


output = []
backtrack()
print(output)
